syntax = "proto3";

package trading;

import "trading/market_data.proto";

service TransformerService {
  rpc Transform(TransformRequest) returns (TransformResponse);
  rpc BulkTransform(BulkTransformRequest) returns (BulkTransformResponse);
}

message TransformRequest {
  FetchResponse raw_quote = 1;
}

message TransformResponse {
  string symbol = 1;
  double price = 2;
  double change = 3;
  double change_percent = 4;
  double open = 5;
  double high = 6;
  double low = 7;
  double volume = 8;
  int64 timestamp = 9;
  string name = 10;
  string exchange = 11;
  string currency = 12;
  double previous_close = 13;
  bool is_market_open = 14;
  double average_volume = 15;
  double fifty_two_week_low = 16;
  double fifty_two_week_high = 17;

  // Computed by transformer
  double day_range_pct = 18;
  double fifty_two_week_pct = 19;
  double gap_pct = 20;
  double volume_ratio = 21;
  double intraday_range_pct = 22;
  string signal = 23;
  string example = 24;
}

message BulkTransformRequest {
  repeated FetchResponse raw_quotes = 1;
}

message BulkTransformResponse {
  repeated TransformResponse quotes = 1;
}
